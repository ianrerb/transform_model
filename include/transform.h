#ifndef TRANSFORM_H
#define TRANSFORM_H
#define PI M_PI 
#include"pricemodel.h"
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>
#include<iostream>
#include<string>

struct Option {double strike; double premium; }; //struct to hold strike, premium pairs

/********************************************
 * ABSTRACT BASE CLASS FOR TRANSFORM CLASSES 
 ********************************************/

class transform_base {//abc
  public:
    transform_base(unsigned int size, bool test = false):N_(size), testmode(test){ };
    virtual ~transform_base(){ };
    
    void N(const unsigned int &val){ N_ = val; };
    unsigned int N() const { return N_; };

    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const =0;

  protected:
    unsigned int N_;
   
    bool testmode;
    void printComplexArray(double complex array[], unsigned int length, std::string title) const {
      std::cout<<"Array: "<<title<<std::endl;
      for(unsigned int i=0; i!=length; i++){
	std::cout<<"("<<creal(array[i])<<","<<cimag(array[i])<<")"<<"  ";
	}
      std::cout<<std::endl;
      };
 };
 
/**********************************
 * BEGIN DEFINITION OF FFT CLASS 
 **********************************/

 class FFT : public transform_base {
  public:
    FFT(unsigned int size, double alpha_,double eta_, bool test = false) : transform_base(size,test), alpha(alpha_), eta(eta_){ }; 
    virtual ~FFT(){ }; 
    
    void Alpha(const double &val){ alpha = val; };
    void Eta(const double &val){eta = val; };
    
    double Alpha() const { return alpha; };
    double Eta() const { return eta; };
    virtual double Lambda() const {return 2.0*PI/(static_cast<double>(N_)*eta); };
    
    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FFTengine.cpp

  protected:
    double alpha, eta;

    void ComputeX(double complex X[], const double &Spot, const double &Strike, const double &C, const pricemodel &model) const;  //defined in FFTengine.cpp
  };

/**********************************
 * BEGIN DEFINITION OF FrFFT CLASS 
 **********************************/

class FrFFT : public FFT {
  public:
  FrFFT(unsigned int size, double alpha_, double eta_, double lambda_, bool test = false): FFT(size, alpha_, eta_,test), lambda(lambda_){};
  ~FrFFT(){ };
  
  void Lambda(const double &l){ lambda = l; };
  double Lambda() const { return lambda; };
  
  double Gamma() const { return eta*lambda/(2.0*PI); };

  std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FrFFT.cpp
  
  private:
    double lambda;
         
    void ComputeY(double complex Y[],const double complex X[]) const; //defined in FrFFT.cpp

    void ComputeZ(double complex Z[]) const; //defined in FrFFT.cpp

    void ComputeXi(double complex Xi[], const double complex Y[], const double complex Z[]) const; //defined in FrFFT.cpp

  };

#endif
