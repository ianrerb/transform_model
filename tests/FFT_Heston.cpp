#include"../include/transform.h"
#include<cmath>
#include<iostream>
#include<vector>

using namespace std;

int main(){
  double Spot, Strike, rate, q, T, sigma, v0, kappa, theta, rho ;
  Spot = 100;
  Strike = 90;
  kappa = 2;
  theta = .04;
  sigma = .5;
  v0=.04;
  rho=-0.7;
  rate = .03;
  q = 0.0;
  T=.5;

  Heston model(kappa, theta, sigma, v0, rho, rate,q,T);
  FFT Engine(0,0,0);

  vector<double> alpha, eta;
  vector<unsigned int> n; 
  
  alpha.push_back(.01);
  alpha.push_back(.5);
  alpha.push_back(1);
  alpha.push_back(1.5);
  alpha.push_back(2);
  alpha.push_back(5);
  alpha.push_back(10);

  eta.push_back(.15);
  eta.push_back(.1);
  eta.push_back(.05);
  
  n.push_back(64);
  n.push_back(256);
  n.push_back(1024);

  for(vector<double>::iterator ait = alpha.begin(); ait!=alpha.end(); ait++){
    for(vector<double>::iterator eit = eta.begin(); eit!= eta.end(); eit++){
      for(vector<unsigned int>::iterator nit = n.begin(); nit!=n.end(); nit++){
	  Engine.setN(*nit);
	  Engine.setAlpha(*ait);
	  Engine.setEta(*eit);
	  cout<<Engine.Prices(Spot,Strike,exp(-rate*T),model)[*nit/2].premium<<"  ";
	}
      }
    cout<<endl;
    }
  return 0;
  }
