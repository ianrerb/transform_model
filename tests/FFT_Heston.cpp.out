203.741  204.212  204.199  126.462  126.355  126.343  55.0927  51.3176  51.7646  
11.3904  11.9991  11.9862  11.8011  12.0045  11.9863  14.2287  11.4703  11.9803  
11.2399  11.9994  11.9862  11.4512  12.0119  11.9863  12.9263  11.4034  11.9798  
11.0814  11.9995  11.9862  11.0724  12.0204  11.9863  11.5811  11.3353  11.9792  
10.9175  11.9996  11.9862  10.6772  12.0301  11.9863  10.3114  11.267  11.9786  
9.97984  11.9956  11.9862  8.52078  12.117  11.9864  5.6845  10.915  11.9745  
9.46203  11.9547  11.9862  7.13356  12.3992  11.9866  3.91099  11.0227  11.9672  
